codeU2.py

```diff
@@ -0,0 +1,181 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Jan 13 10:22:28 2017
+Exercise set 2, Code U
+@author: Luiza
+"""
+import logging
+import unittest
+
+class Tree():
+    
+    """Defines the class for the tree:
+    self.children contains a dict of pairs (node:(left_child,right_child))
+    self.head contains the value of the tree head
+    None denotes the absence if the node
+    """
+    
+    def __init__(self,children,head):
+        self.children = children;
+        self.head = head;
+
+def dfs_path(tree_, start, end, path=None):
+
+    """Yields all full paths from start node
+    to end node, excluding the end node
+    """
+    
+    tree = tree_.children
+    stack = [(start, [start])]
+    while stack:
+        (vertex, path) = stack.pop()
+        if vertex is None:
+            continue
+        for next in set(tree[vertex]) - set(path):
+            if next == end:
+                yield path
+            else:
+                stack.append((next, path+[next]))
+
+def get_ancestors(tree_, node):
+
+    """Returns the ancestors from head to a given node
+    """
+
+    head = tree_.head
+    if not head:
+        logging.warning('tree is empty, no ancestors')
+        return None
+    if head == node:
+        logging.warning('Head has no ancestors')
+        return None
+    ancestors = list(dfs_path(tree_, head, node, path=None))
+    if ancestors:
+        return ancestors[0]
+    else:
+        assert ValueError, 'Node ' + str(node) + ' does not exist in the tree'
+
+def get_least_common_ancestor(tree_, node1, node2):
+
+    """Returns the least common ancestor
+    for two given nodes
+    """
+
+    tree = tree_.children
+    head = tree_.head
+    if not tree:
```

Comment:
Likewise for these checks. Some of them duplicate checks that are done above (`node1 == head`) and the checking could probably be simplified. Could/should `get_ancestors` return an empty `list` (`[]`) instead of `None` for that case? It sounds like it would make sense and then the algorithm would just work for `node == head` case as well without any extra checks required.